feat(nuxt): route announcer

ivan-kalachikov opened the pull request:
<!---
☝️ PR title should follow conventional commits (https://conventionalcommits.org)

Please carefully read the contribution docs before creating a pull request
 👉 https://nuxt.com/docs/community/contribution
-->

### 🔗 Linked issue
[14673](https://github.com/nuxt/nuxt/issues/14673)
<!-- Please ensure there is an open issue and mention its number as #123 -->

### ❓ Type of change

<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->

- [ ] 📖 Documentation (updates to the documentation, readme or JSdoc annotations)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [ ] 👌 Enhancement (improving an existing functionality like performance)
- [x] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description
Resolves [#14673](https://github.com/nuxt/nuxt/issues/14673)

**Adding Route Announcer to Nuxt js for better accessibility.**

- As reference was taken: [https://github.com/vue-a11y/vue-announcer](https://github.com/vue-a11y/vue-announcer) 
- Implemented as a built-in component, as Daniel suggested here: [https://github.com/nuxt/nuxt/issues/23255#issuecomment-1732288389](https://github.com/nuxt/nuxt/issues/23255#issuecomment-1732288389)
<!-- Describe your changes in detail -->
<!-- Why is this change required? What problem does it solve? -->
<!-- If it resolves an open issue, please link to the issue here. For example "Resolves #1337" -->

### 📝 Checklist

<!-- Put an `x` in all the boxes that apply. -->
<!-- If your change requires a documentation PR, please link it appropriately -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have linked an issue or discussion.
- [x] I have added tests (if possible).
- [x] I have updated the documentation accordingly.

ivan-kalachikov committed:
feat(nuxt): add route announcer compasable and component

ivan-kalachikov commented:
TODO:
[ ] Handle route change during async page loading
[ ] Cleanup
[ ] Write documentation
[ ] Write tests

ivan-kalachikov committed:
feat(nuxt): add route announcer compasable and component

ivan-kalachikov committed:
remove unnessessary changes and fix types

ivan-kalachikov committed:
add mode client

ivan-kalachikov committed:
add a draft of documentation

ivan-kalachikov committed:
update documentation

ivan-kalachikov committed:
refactor composable and component

ivan-kalachikov committed:
return mistakenly deleted line

ivan-kalachikov committed:
fix documentation

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

ivan-kalachikov committed:
fix path

ivan-kalachikov committed:
Merge branch 'feat/route-announcer' of github.com:ivan-kalachikov/nuxt into feat/route-announcer

ivan-kalachikov committed:
Merge branch 'main' of github.com:ivan-kalachikov/nuxt into feat/route-announcer

ivan-kalachikov committed:
Merge branch 'main' into feat/route-announcer

ivan-kalachikov committed:
update documentation

ivan-kalachikov committed:
refactoring the composable and the component

ivan-kalachikov committed:
add test case

ivan-kalachikov committed:
Merge branch 'main' of github.com:ivan-kalachikov/nuxt into feat/route-announcer

ivan-kalachikov committed:
remove unused imports

ivan-kalachikov committed:
fix linter

ivan-kalachikov committed:
fix the test case

ivan-kalachikov committed:
Merge branch 'nuxt:main' into feat/route-announcer

ivan-kalachikov committed:
fix linter

ivan-kalachikov committed:
minor refactoring

ivan-kalachikov committed:
remove unnecessary changes

ivan-kalachikov committed:
add a test

ivan-kalachikov committed:
Merge branch 'nuxt:main' into feat/route-announcer

ivan-kalachikov committed:
Merge branch 'feat/route-announcer' of github.com:ivan-kalachikov/nuxt into feat/route-announcer

ivan-kalachikov committed:
fix test

ivan-kalachikov committed:
refactor composable and tests

ivan-kalachikov committed:
remove test case only

ivan-kalachikov committed:
fix test

ivan-kalachikov committed:
fix test

ivan-kalachikov committed:
fix test and restore playground app page

ivan-kalachikov committed:
fix test

ivan-kalachikov committed:
fix test

ivan-kalachikov committed:
add NuxtLoadingAnnouncer to the playground app.vue

ivan-kalachikov committed:
refactor and fix documentation

ivan-kalachikov committed:
change web component and shadow dom element to regular div due to unresolved issue with ChromeVox screen reader

ivan-kalachikov committed:
change waiting for timeout to waiting for request animation frame

ivan-kalachikov committed:
Merge branch 'nuxt:main' into feat/route-announcer

ivan-kalachikov committed:
Merge branch 'feat/route-announcer' of github.com:ivan-kalachikov/nuxt into feat/route-announcer

ivan-kalachikov committed:
update documentation

ivan-kalachikov committed:
remove announcer from playground app.vue

ivan-kalachikov committed:
minor refactoring

danielroe commented on packages/nuxt/src/app/composables/route-announcer.ts:
```diff
@@ -0,0 +1,102 @@
+  }
+
+  function _updateMessageWithPageHeading () {
+    set(document?.title?.trim(), politeness.value)
```
this might be tricky as the title might not be updated synchronously

cc: @harlan-zw - is there an unhead hook we could use, or can you think of a better way to implement?

ivan-kalachikov replied:
@danielroe thanks for the idea, Unhead hook `dom:rendered` looks much more suitable indeed.

Let see what @harlan-zw will answer, I'll push the commit with refactoring using `dom:rendered` meanwhile

ivan-kalachikov replied:
✅ Refactored with `dom:rendered` hook

ivan-kalachikov committed:
refactor with Unhead hook and update tests

ivan-kalachikov committed:
Merge branch 'nuxt:main' into feat/route-announcer

ivan-kalachikov committed:
update documentation

ivan-kalachikov committed:
Merge branch 'main' into feat/route-announcer

ivan-kalachikov committed:
Merge branch 'main' into feat/route-announcer

danielroe commented on packages/nuxt/src/app/composables/route-announcer.ts:
```diff
@@ -0,0 +1,94 @@
+
+  _updateMessageWithPageHeading()
+
+  activeHead?.hooks?.hook('dom:rendered', () => {
+    _updateMessageWithPageHeading()
+  })
```
maybe we could unregister this hook within `onScopeDispose`

danielroe commented on packages/nuxt/src/app/components/nuxt-route-announcer.ts:
```diff
@@ -0,0 +1,49 @@
+      set, polite, assertive, message, politeness
+    })
+
+    return () => h('nuxt-route-announcer', {
+      class: 'nuxt-route-announcer',
+      style: {
+        position: 'absolute'
+      }
+    }, h('div', {
```
do we need to render the `h('nuxt-route-announcer')`? Couldn't we just render the div?

ivan-kalachikov replied:
Not really, you are right it's not necessary. Will change it in a half an hour.
It is an artifact of my initial implementation with web-component `nuxt-route-announcer`

ivan-kalachikov replied:
`nuxt-route-announcer` replaced with `div`  ✅

ivan-kalachikov committed:
fix readme

ivan-kalachikov committed:
replace nuxt-route-announcer tag with div

ivan-kalachikov committed:
Merge branch 'main' into feat/route-announcer

danielroe commented:
Beautiful work.

ivan-kalachikov committed:
Merge branch 'main' into feat/route-announcer

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

danielroe reviewed (commented):
We need to add a `@since` modifier to the composables (3.12.0) and add a comment in the docs that this will be available in v3.12.

harlan-zw commented on packages/nuxt/src/app/composables/route-announcer.ts:
```diff
@@ -0,0 +1,94 @@
+function createRouteAnnouncer (opts: NuxtRouteAnnouncerOpts = {}) {
+  const message = ref('')
+  const politeness = ref<PolitenessValue>(opts.politeness || Politeness.Polite)
+  const activeHead = getActiveHead()
```
This needs to come from @unhead/vue to work outside of a Vue context. `injectHead` is the function

ivan-kalachikov replied:
Thanks for pointing this out! I've changed the code to use `injectHead` from `@unhead/vue `

ivan-kalachikov committed:
add @since tag to composable jsdoc

ivan-kalachikov committed:
use proper function to get access to active head

ivan-kalachikov committed:
add note about compasable availability

ivan-kalachikov committed:
add new badge and info about release version to composable and component docs

ivan-kalachikov committed:
fix doc linter

ivan-kalachikov commented:
> We need to add a `@since` modifier to the composables (3.12.0) and add a comment in the docs that this will be available in v3.12.

Added, also added badges `new` to the navigation.

lihbr reviewed (commented):
Love it ☺️

I'm wondering if we could include it in the default layout/root if not present so that Nuxt apps are accessible by default, any thoughts? 🤔

ivan-kalachikov commented:
> I'm wondering if we could include it in the default layout/root if not present so that Nuxt apps are accessible by default, any thoughts? 🤔

Absolutely! Opened an issue in the `nuxt/starter` repository. Will be taken after `3.12.0` release. If nobody minds.

[#716 Add NuxtRouteAnnouncer in a default templates (v3 - *) after Nuxt 3.12.0 release](https://github.com/nuxt/starter/issues/716)

danielroe commented:
I think it should also be in the default app.vue injected by the pages module when one doesn’t already exist. (In this repo.)

danielroe committed:
feat: add nuxt route announcer to default route-based `app.vue`

danielroe committed:
Merge remote-tracking branch 'origin/main' into feat/route-announcer

ivan-kalachikov committed:
fix tests

danielroe commented:
Sorry for breaking the tests! 🙈

ivan-kalachikov committed:
fix tests

ivan-kalachikov committed:
Merge branch 'main' into feat/route-announcer

antfu committed:
docs: update the code style page (#26792)

iiio2 committed:
chore: fix `devcontainer.json` syntax (#26776)

huang-julien committed:
docs: reference correct composable in `callWithNuxt` (#26771)

renovate[bot] committed:
chore(deps): update devdependency @nuxt/test-utils to v3.12.1 (main) (#26793)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

mehdi-coding committed:
docs: add sample app.vue (#26796)

shinGangan committed:
docs: remove direct link to subheading (#26775)

renovate[bot] committed:
chore(deps): update pnpm to v9 (main) (#26799)

kawadah committed:
docs: fix `srcDir` description mentioning deprecated `static/` directory (#26804)

edwinfadul committed:
docs: provide example of passing props to a child page (#26805)

edwinfadul committed:
docs: provide example of exposing a `pageRef` from a child page (#26806)

renovate[bot] committed:
chore(deps): update all non-major dependencies (main) (#26768)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>
Co-authored-by: Daniel Roe <daniel@roe.dev>

danielroe committed:
refactor: remove `Politeness` enum

danielroe committed:
chore: render span to avoid potential hydration errors

danielroe committed:
chore: remove unused import

danielroe committed:
Merge remote-tracking branch 'origin/main' into feat/route-announcer

danielroe committed:
Revert "fix tests"

This reverts commit 5a3dde1e017f9bab2798fca1f97c9980bc07ccf9.

danielroe committed:
Revert "fix tests"

This reverts commit 262448ac3c38203543424c18114aa268ba050d6f.

danielroe committed:
Revert "feat: add nuxt route announcer to default route-based `app.vue`"

This reverts commit fdce83e52c5153af5dcf3c03cbe0485ed57fa214.

danielroe committed:
test: remove .only

danielroe commented:
Looking into it, there seems to be an issue with having anything _but_ a suspense node as root of `app.vue` when it comes to client-side components in current implementation. they are not being rendered by the time the server response is coming. it's not an issue introduced by this PR so I think we can go ahead and merge and investigate later.

cc: @huang-julien

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

danielroe committed:
docs: remove links to nonexistent blog

danielroe commented:
Thank you - this is fantastic! ❤️
